### ISSUE TITLE
Product page refactoring for theme 2.0

### ISSUE THREAD

**Issue by pkarw** [DivanteLtd/vue-storefront#3093]
This is the initial plan to refactor the `Product.vue` page accordingly to #2696 and after the initial learnings we got from the `Category.vue` refactoring (#2890)

The goal is to prepare the `Product.vue` for Theme 2.0 - based on Storefront UI but without dramatic changes, keeping the business logic as is (it's very important assumption) and share it between `product` and `product-next` modules. The same time: improving the Vuex readability, preparing the Product page data interface for the 2.0

Stage 1: to be finished till end of July (preferred date: 3rd week of July)

- [ ] Create the `product-next` module for the consistency with `category-next` - use only those new one’s in the new Product.vue and Category.vue
- [ ] We should re-use as much the original logic of `product` module as we can: via helpers or just via importing the actions directly from the old module (doing that way we don’t have to create additional dependencies and can just do like: `import { list } from `@vue-storefront/store/modules/product/actions`
- [ ] Add suitable getters with the proper naming convention -and don’t use the state directly in the theme (just only thru getters)
- [ ] We don’t touch the business logic: `product/helpers` and the actions like `setupBreadcrumbs`, `setupAssociated`, actions related with bundle etc. Maybe just some minor refactoring like:
    - [ ] avoiding state mutations directly in actions (move to mutations)
    - [ ] remove `rootStore` references,
    - [ ] limit the logic based on events
- Don’t change the state much - to keep the single code base with product 1.0 module as far as we can:
    - [ ] we must keep `current`, `original`, `parent`
    - [ ] we must keep `current_bundle_options`, `current_custom_options`, `custom_options_validators`
    - [ ] we can remove (from both - original and new module): `productLoadStart`, `productLoadPromise`, `productGallery` (or if not, rename it to keep the convention which is `under_scored`)
    - [ ] remove `color` and `size` from `current_options` - change the theme to not access this property ditrectly but via getter
    - [ ] we can remove `list` -> but we must refactor the action `product/list` (extract to helper which returns the data + mutate the state after helper is executed in the original `product/list`; in the `product-next/list` (or whatever different name it will have) we can remove this state property as the list ist now in the category)
- Rename getters to the proper naming convention:
    - [ ] `getParentProduct`,
    - [ ] `getCurrentProduct`
    - [ ] `getCurrentProductConfiguration`
    - [ ] `getOriginalProduct`
    - [ ] `getCurrentProductOptions`
    - [ ] `getCurrentProductGallery`
    - [ ] `getCurrenntProductBreadcrumbs`
    - ...
- Refactor Product.vue + remove the `core/pages/Product.js` dependency
    - [ ] map getters with the right getters- https://github.com/DivanteLtd/vue-storefront/blob/70543b3cfa379971220f6514348fe0386a108878/core/pages/Product.js#L25
    - [ ] `attributesByCode` and `attributesByid` seems to not being used anymore and can be removed
    - [ ] shall we have `this.product` or `this.getProduct` (as getter suggest)?
    - [ ] `productName`, `productId` - should be probable replaced with just `product.name` and `productid` in the theme
    - [ ] `customAttributes` ->replaced with vex getter `getCurrentProductCustomAttributes` or just `getCustomAttributes` mapped to `custom_attributes` (as it’s related to the page context only)
    - [ ] Wishlist and Compare should be moved to smart components - the one’s that uses the Vue to perform the logic, but product page shouldn’t be aware of them at all
    - [ ] `isOptionAvailable` could be moved to helper (called just from theme) or can stay here
    - [ ] `onAfterCustomOptionsChanged`, `onAfterBundleOptionsChanged` logic - to be moved to Vuex handler
    - [ ] `onStateCheck` to be removed - the parent check moved to the Vuex
    - [ ] check when `onAfterPriceUpdate` is called (tier prices probably) and check if we can remove it (probably we can move it to vuex + modify tier prices to call this vex action instead)
    - [ ] Not sure if we can move `onAfterFilterChanged` to vex as it replaces parts of url but probably we can
    - [ ] Remove `onUserPricesRefreshed` - I mean call `asyncData` or other vex method instead
- Refactor the `product/actions` and `product/helpers`
    - [ ] replace `state` read operations with getters
    - [ ] replace `state` write operations with mutations
    - [ ] Extract network call from `syncPlatformPricesOver` to helper and then data resolver - remove the `callback_event` (it’s not used)
    - [ ] In the new module `product-next` remove the `setCurrentProduct` and `selectDefaultVariant` from `product-next/single` or its equivalent - the logic of `product-single` should be shared between old and new product modules helper or import; old action should emulate these parameters new one - should just return product data (setting current product should be in different vex action - related to the product page only)
    - [ ] `product/list` should not `updateState` - i mean the new version should not update the state, just return the final data + do caching the `category-next` current call to `product/list` should be replaced with the new method
    - [ ] `related`, `setCurrentOption`, `setCustomOptions`  and `setCurrentErrors` are not used and can be removed
    - [ ] `fetch` and `fetchAsync` should be merged into one call `loadProduct` or `fetchProduct` that will be used in the `Product.vue`

**Comment by pkarw** [comment]
@filrak, @patzick, @lukeromanowicz - this one is for discussion next week and to start right after #2696 (from 1st of July probably) and hopefully, finish it in the 3rd week of July

**Comment by pkarw** [comment]
update: `setupBreadcrumbs` can be re-used from the `category-next` after refactor @patzick

**Comment by filrak** [comment]
> shall we have this.product or this.getProduct (as getter suggest)?
`this.product` -> getter should get product data and assign to `product` local var.

Also shouldn't we have a general breadcrumbs module to not limit them to product/cat? I think it'd be better for SEO

**Comment by filrak** [comment]
`getParentProduct` & `getOriginalProduct` seem very similar and not straightforward to understand what those getters are holding.

**Comment by GabiDivante** [comment]
@patzick I've retested everything in #3217 and the product page overall:
- regarding issues which you couldn't reproduce - I couldn't as well, so I think these no longer occur,
- all fixes work fine,
- I just found one problem related to #2733 - in the offline mode, we cannot add any product to the cart due to hanged availability check - there are loading circle and inactive button visible till we go back to the online. Should I reopen #2733? 
![Selection_073](https://user-images.githubusercontent.com/49188852/65048850-16dcbc80-d965-11e9-815d-6ef48c95e416.png)

**Comment by pkarw** [comment]
Thanks @GabiDivante it works just fine on demo: https://demo.storefrontcloud.io/gear/gear-3/sprite-yoga-companion-kit-45.html

**Comment by alinadivante** [comment]
@pkarw I changed my endpoints to `https://demo.storefrontcloud.io/api/` but I also have problem with addin a bundle product to the cart.
Error from dev console:
`mutations.ts:100 Deprecated mutation CATALOG_UPD_BUNDLE_OPTION - use PRODUCT_SET_BUNDLE_OPTION instead`

**Comment by pkarw** [comment]
@alinadivante ok cool, so this will be easy to fix - @patzick
